v23.1/transaction-retry-error-reference.md

```diff
@@ -1,54 +1,92 @@
+#### Client-side retry handling example
 
-## Error reference
+For a conceptual example of application-defined retry logic, and testing that logic against your application's needs, see the [client-side retry handling example](transaction-retry-error-example.html).
+
+### Minimize transaction retry errors
+
+In addition to the steps described in [Client-side retry handling](#client-side-retry-handling), which detail how to configure your application to restart a failed transaction, there are also a number of changes you can make to your application logic to prevent client-side transaction retry errors from occurring in the first place:
+
+1. Limit the number of affected rows by following [performance-tuning best practices](apply-statement-performance-rules.html) (e.g., query performance tuning, index design and maintenance, etc.). Not only will transactions run faster and hold locks for a shorter duration, but the chances of [read invalidation](architecture/transaction-layer.html#read-refreshing) when the transaction’s [timestamp is pushed](architecture/transaction-layer.html#timestamp-cache) due to a conflicting write is decreased due to a smaller read set (i.e., a smaller number of rows read).
+
+1. Break down larger transactions into smaller ones (e.g., [bulk deletes](bulk-delete-data.html)) to have transactions hold locks for a shorter duration. This will also decrease the likelihood of [pushed timestamps](architecture/transaction-layer.html#timestamp-cache) and retry errors. For instance, as the size of writes (number of rows written) decreases, the chances of the (bulk delete) transaction’s timestamp getting bumped by concurrent reads decreases.
+
+1. Design your applications to reduce network round trips by [sending statements in transactions as a single batch](transactions.html#batched-statements) (e.g., using [common table expressions](common-table-expressions.html)). Batching allows CockroachDB to [automatically retry](transactions.html#automatic-retries) a transaction when [previous reads are invalidated](architecture/transaction-layer.html#read-refreshing) at a [pushed timestamp](architecture/transaction-layer.html#timestamp-cache).
+
+1. Limit the size of the result sets of your transactions so that CockroachDB is more likely to [automatically retry](transactions.html#automatic-retries) when [previous reads are invalidated](architecture/transaction-layer.html#read-refreshing) at a [pushed timestamp](architecture/transaction-layer.html#timestamp-cache).
```

Comment:
Maybe I just don't understand database stuff? I never work with SQL logic over in security land. But I don't at all follow this "...so that..."

Also, isn't this recommendation (limit the size of result sets) at odds with the previous one (batch transactions together), insofar as batched transactions will often have larger result sets?

Reply:
You're right that these two sound about the same, but I was asked to make them explicit separate guidance. I'll ask about this specifically to be sure. Thanks!

Reply:
Overall goal is to allow CRDB to automatically retry. If user does not batch, no auto-retry. If user batches, but resultset hits 16KB, no auto-retry. If batch, and resultset is <16KB, 👍

But yeah, they are related since batching will increase the resultset.